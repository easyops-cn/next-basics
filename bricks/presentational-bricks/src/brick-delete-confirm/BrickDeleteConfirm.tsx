import React, { ChangeEvent, useState } from "react";
import { Input, Modal } from "antd";
import { ButtonProps } from "antd/lib/button";
import style from "./index.module.css";

export interface LegacyBrickDeleteConfirmProps {
  deleteName: string;
  originBtnProps?: ButtonProps;
  onDelete: () => void;
  className?: any;
  visible?: boolean;
  loading: boolean;
  handleCancel?: () => void;
}

export interface CardBodyProps {
  name: string;
  content: string;
  handleChange: (value: string) => void;
}

export interface CardBodyState {
  value: string;
}

const CardTitle = ({ name }: { name: string }) => {
  return (
    <div>
      确认要删除 <span className={style.highLight}>{name}</span> 吗？
    </div>
  );
};
export function CardBody(props: CardBodyProps): React.ReactElement {
  const [value, setValue] = useState(props.content);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setValue(e.target.value);
    props.handleChange(e.target.value);
  };

  return (
    <div>
      <div style={{ marginBottom: 15 }}>
        该删除操作将抹除所有与其相关的数据，请确认后在下方输入{" "}
        <span className={style.highLight}>{props.name}</span> 来解锁确定按钮。
      </div>
      <Input
        name="deleteName"
        size="small"
        value={value}
        onChange={handleChange}
      />
    </div>
  );
}

export function BrickDeleteConfirm(
  props: LegacyBrickDeleteConfirmProps
): React.ReactElement {
  const [disabled, setDisabled] = useState(true);
  const [content, setContent] = useState("");

  const handleChange = (value: string) => {
    setContent(value);
    setDisabled(value !== props.deleteName);
  };

  const handleOk = () => {
    props.onDelete && props.onDelete();
  };

  const handleCancel = () => {
    setDisabled(true);
    setContent("");
    props.handleCancel();
  };

  return (
    <>
      <Modal
        title={<CardTitle name={props.deleteName} />}
        visible={props.visible}
        onOk={handleOk}
        onCancel={handleCancel}
        okText="确定"
        cancelText="取消"
        okType="danger"
        okButtonProps={{ disabled: disabled, loading: props.loading }}
        destroyOnClose={true}
        maskClosable={false}
        width={416}
      >
        {
          <CardBody
            name={props.deleteName}
            content={content}
            handleChange={value => handleChange(value)}
          />
        }
      </Modal>
    </>
  );
}
