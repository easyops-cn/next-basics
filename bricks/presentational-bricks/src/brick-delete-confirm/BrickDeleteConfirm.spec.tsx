import React from "react";
import { shallow, mount } from "enzyme";
import { BrickDeleteConfirm, CardBody } from "./BrickDeleteConfirm";

describe("BrickDeleteConfirm", () => {
  const deleteName = "console-w";

  const onDelete = jest.fn();
  const handleCancel = jest.fn();

  it("should work", () => {
    const wrapper = shallow(
      <BrickDeleteConfirm
        deleteName={deleteName}
        onDelete={onDelete}
        handleCancel={handleCancel}
        visible={true}
        loading={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should trigger onDelete when modal is ok", () => {
    const mockOnDelete = jest.fn();
    const wrapper = mount(
      <BrickDeleteConfirm
        deleteName={deleteName}
        onDelete={mockOnDelete}
        handleCancel={handleCancel}
        visible={true}
        loading={true}
      />
    );
    wrapper.find("Modal").invoke("onOk")();
    expect(mockOnDelete).toBeCalled();
    wrapper.find("Modal").invoke("onCancel")();
    expect(wrapper.find("CardBody").prop("content")).toBe("");
    wrapper.find("CardBody").invoke("handleChange")("123");
    expect(wrapper.find("CardBody").prop("content")).toBe("123");
  });

  it("should change the value", () => {
    const handleChange = jest.fn();

    const wrapper = mount(
      <CardBody name={deleteName} content="" handleChange={handleChange} />
    );

    wrapper
      .find("input")
      .simulate("change", { target: { value: "console-W" } });

    expect(handleChange).toHaveBeenCalled();
    expect(wrapper.find("input").props().value).toEqual("console-W");
  });
});
