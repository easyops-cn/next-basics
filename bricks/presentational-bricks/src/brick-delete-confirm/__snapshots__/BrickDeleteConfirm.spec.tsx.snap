// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickDeleteConfirm should work 1`] = `
<Fragment>
  <Modal
    cancelText="取消"
    confirmLoading={false}
    destroyOnClose={true}
    maskClosable={false}
    maskTransitionName="fade"
    okButtonProps={
      Object {
        "disabled": true,
        "loading": false,
      }
    }
    okText="确定"
    okType="danger"
    onCancel={[Function]}
    onOk={[Function]}
    title={
      <CardTitle
        name="console-w"
      />
    }
    transitionName="zoom"
    visible={true}
    width={416}
  >
    <CardBody
      content=""
      handleChange={[Function]}
      name="console-w"
    />
  </Modal>
</Fragment>
`;
